### PR TITLE
Add wrappers for old value and old property in BatchMaterial

### LINKED ISSUE #25056: Support old variable value and old material property in BatchMaterial
## Reason

The batch computation may sometimes depend on the old variable value and old material property. Currently we don't have wrappers for them in BatchMaterial.

## Design

Add the missing wrappers.

### PR DESCRIPTION
**Pull request by hugary1995**
close #25056

### CODE REVIEW

**Comment by hugary1995** [comment]
@dschwen This is in support of NEML2 integration into blackbear which I wish to complete by the end of August before I send the report for internal review.

**Review by dschwen — DISMISSED** [review]
Looks good

File: framework/include/userobjects/BatchMaterial.h [hunk 1]
```diff
@@ -35,6 +35,19 @@ struct GatherMatProp
...
+template <typename T>
+struct GatherMatPropOld
+{
+  typedef T type;
+  typedef const MaterialProperty<T> gather_type;
+
+  template <typename M>
+  static gather_type * getPointer(M & mpi, const MaterialPropertyName & name)
+  {
+    return &mpi.template getMaterialPropertyOld<T>(name);
```
Comment (lindsayad) [inline framework/include/userobjects/BatchMaterial.h:47]: After https://github.com/idaholab/moose/pull/24673, `getMaterialProperty` takes a state argument with 0 denoting current, 1 denoting old, etc. So if you wanted to slightly reduce code, you could add add an `unsigned int` template argument to the existing `GatherMatProp` struct

Comment (lindsayad) [reply framework/include/userobjects/BatchMaterial.h:47]: And by doing so you'd immediately enable older properties

**Review by lindsayad — CHANGES_REQUESTED** [review]
see comment

File: framework/include/userobjects/BatchMaterial.h (hunk 1 above)
Comment (hugary1995) [reply framework/include/userobjects/BatchMaterial.h]: That's a cool enhancement! For now I only enabled GatherMatPropTempl<T, 0> and GatherMatPropTempl<T, 1>. I'll leave the other cases (older state, AD) and their regression tests to whoever needs them.

**Review by dschwen — APPROVED** [review]

**Review by lindsayad — APPROVED** [review]

**Pull request merged**